### ISSUE TITLE
How the client to process incoming messages in parallel?

### ISSUE THREAD

**Issue by xzzh999** [nats-io/nats.go#466]
Hi~
Great work!

As the title said, should I just ```make multiple clients``` to do this?  Is there a convenient wrapped way?

Need u help, thx~

**Comment by ripienaar** [comment]
Multiple is fine. In my experience one connection can handle thousands of messages a second. 

You make one thread that receives messages and put them in an internal work queue and other thread or a pool of threads to process that queue.

**Comment by xzzh999** [comment]
@ripienaar 
Thank you for u reply~

In some cases, for example: ```a searching task```, it will always ```take 6 seconds to finish```, and when the task not done, there is another task comming, and i want to process the follow task as quick as possible, that means i want ```the handle function called in parallel```, not just one by one.

At the [nats.io website](https://nats.io/documentation/writing_applications/advanced/) document advanced page at the ```Slow Consumers``` part,  it not tell us how to deal with this. 

If i use a queue to cache the messages, can i call ```the same client``` to reply? ```Is that thread safe``` when i use ```multithread mode``` of goroutines?

**Comment by ripienaar** [comment]
Yes it’s thread safe. 

The routine receiving the messages from NATS must not block or get delayed else it will disconnect. Even if you did not want parallel processing 2 routines is the way to go ime

**Comment by xzzh999** [comment]
@ripienaar 
Not one productor one consumer but ```one productor multiple consumers``` to access one client, it is also thread-safe?

**Comment by ripienaar** [comment]
Yes should be fine

**Comment by xzzh999** [comment]
@ripienaar Thanks~

**Comment by derekcollison** [comment]
Which client language are you using?

**Comment by xzzh999** [comment]
@derekcollison 

I use golang

**Comment by derekcollison** [comment]
ok so if the time it takes to process a message is long, like RI said, I would propose the following, again assuming you were not required to process serially.

Create a buffered channel. Have the subscriber callback simply place the message or unmarshalled struct onto the queue. Have some number of Go routines spin up and pull from the channel, process the work and send directly back from the Go routine. This is totally safe with the Go client.

**Comment by xzzh999** [comment]
@derekcollison 

Very detailed and sounds perfect， i will try that,  thank you~

**Comment by derekcollison** [comment]
Happy to help. If your processing is CPU bound good rule of thumb is number of Go routines == number of cores. If more IO bound you can run a higher level of Go routines. If you know what the expected inbound rate of requests are you can also use that to plan SLA's per application and number of applications you may want to run in parallel.

**Comment by xzzh999** [comment]
@derekcollison 

Got it, multiple processes is a good suggestion to guaranteed reliability, and  i want both:  one application process multiple messages in parallel, and use multiple processes in the system  just like the microservice does.

**Comment by chenjpu** [comment]
@derekcollison @ripienaar When the message is processed with asynchronous tasks, the stats are distorted, and the stats are the time to call the asynchronous task, not the message processing time

**Comment by derekcollison** [comment]
Which stats?

**Comment by chenjpu** [comment]
[service endpoint  stats](https://github.com/nats-io/nats.go/blob/46be7408f6630569929044fcad459ba0844a296d/micro/service.go#L648)

**Comment by derekcollison** [comment]
Why is that the case?

**Comment by chenjpu** [comment]
Using request-reply to implement the service request, in the response function, using go pool improves the throughput, because each processing takes 10 to 100 ms of time.
